tools/tpm2_pcrallocate.c

```diff
@@ -0,0 +1,241 @@
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//**********************************************************************;
+
+#include <ctype.h>
+#include <errno.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <tss2/tss2_esys.h>
+
+#include "log.h"
+#include "pcr.h"
+#include "tpm2_options.h"
+#include "tpm2_tool.h"
+#include "tpm2_util.h"
+#include "tpm2_alg_util.h"
+
+struct {
+    TPM2B_AUTH platauth;
+    TPML_PCR_SELECTION pcrSelection;
+} ctx;
+
+
+/*
+ * Dump the TPML_PCR_SELECTION
+ */
+static void
+dump_pcrselection (TPMS_PCR_SELECTION   pcrselection[],
+                   UINT32               count)
+{
+    UINT32 i;
+    int j;
+    bool first;
+    tpm2_tool_output ("Selected PCRs:\n");
+    for (i = 0; i < count; i++) {
+        switch(pcrselection[i].hash) {
+            case TPM2_ALG_SHA1:
+                tpm2_tool_output ("  - bank: sha1\n");
+                break;
+            case TPM2_ALG_SHA256:
+                tpm2_tool_output ("  - bank: sha256\n");
+                break;
+            case TPM2_ALG_SHA384:
+                tpm2_tool_output ("  - bank: sha384\n");
+                break;
+            case TPM2_ALG_SHA512:
+                tpm2_tool_output ("  - bank: sha512\n");
+                break;
+            default:
+                tpm2_tool_output ("  - bank: alg0x%08x\n",
+                                  pcrselection[i].hash);
+        }
+        first = true;
+        for (j = 0; j < pcrselection[i].sizeofSelect * 8; j++) {
+            if (pcrselection[i].pcrSelect[j / 8] & 1<<(j % 8)) {
+                if (first) {
+                    tpm2_tool_output ("    pcrs: [ %i", j);
+                    first = false;
+                } else {
+                    tpm2_tool_output(", %i", j);
+                }
+            }
+        }
+        tpm2_tool_output (" ]\n");
+    }
+}
+
+static bool pcr_allocate(ESYS_CONTEXT *ectx) {
+    TSS2_RC rval;
+    TPMI_YES_NO allocationSuccess;
+    UINT32 maxPCR;
+    UINT32 sizeNeeded;
+    UINT32 sizeAvailable;
+
+    rval = Esys_TR_SetAuth(ectx, ESYS_TR_RH_PLATFORM, &ctx.platauth);
+    if (rval != TSS2_RC_SUCCESS) {
+        LOG_ERR("Could not set the password.");
+        LOG_PERR(Esys_TR_SetAuth, rval);
+        return false;
+    }
+
+    dump_pcrselection(ctx.pcrSelection.pcrSelections, ctx.pcrSelection.count);
+
+    rval = Esys_PCR_Allocate(ectx, ESYS_TR_RH_PLATFORM, 
+                             ESYS_TR_PASSWORD, ESYS_TR_NONE, ESYS_TR_NONE,
+                             &ctx.pcrSelection, &allocationSuccess,
+                             &maxPCR, &sizeNeeded, &sizeAvailable);
+    if (rval != TSS2_RC_SUCCESS) {
```

Comment:
`maxPCR` could be printted here to inform the user on the maximum PCR selection possible in any case.

Reply:
Unfortunately, maxPCRs is only returned if the program succeeds...

Reply:
Unfortunately maxPCR is only set if rval == TSS2_RC_SUCCESS, but added maxPCR to allocationSuccess != YES...